Reapply "kernel: modules: fix led-group-multi-color patch"

professor-jonny opened the pull request:
Reapply "kernel: modules: fix led-group-multi-color patch"
This reverts commit https://github.com/openwrt/openwrt/commit/1375a7bae0d5e8dddb6b6cfe0584a399cd6c20b4.

Fixes: https://github.com/openwrt/openwrt/commit/5fd86d66c15fc4d6f5076f7d07c91e0690d91531 ("kernel: modules: add support for led-group-multicolor")

Signed-off-by: jonathan brophy <professor_jonny@hotmail.com>

professor-jonny committed:
kernel: modules: reenable led-group-multi-color patch

This reverts commit 1375a7bae0d5e8dddb6b6cfe0584a399cd6c20b4.

This version does not activate the CONFIG_LEDS_CLASS_MULTICOLOR option,
this option is controlled by the generic target configuration.

Signed-off-by: jonathan brophy <professor_jonny@hotmail.com>
Link: https://github.com/openwrt/openwrt/pull/17277
[Reword commit message]
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented on package/kernel/linux/modules/leds.mk:
```diff
@@ -24,6 +24,24 @@ $(eval $(call KernelPackage,leds-gpio))
+  TITLE:=LEDs group multi-color support
+  KCONFIG:=CONFIG_LEDS_CLASS_MULTICOLOR \
+     CONFIG_LEDS_GROUP_MULTICOLOR
+  FILES:=$(LINUX_DIR)/drivers/leds/rgb/leds-group-multicolor.ko
```
```
FILES:=$(LINUX_DIR)/drivers/leds/rgb/leds-group-multicolor.ko
     $(LINUX_DIR)/drivers/leds/led-class-multicolor.ko
```

professor-jonny replied:
Would it make sense to add that as a separate module and add a depends ?

hauke replied:
`CONFIG_LEDS_CLASS_MULTICOLOR` is always activated except for the stm32 target.

I think it is best to fix the stm32 target to not activate it any more.

professor-jonny replied:
You mean remove it by default in every platform and add it as a depends to every kmod that requires it?

hauke replied:
`CONFIG_LEDS_CLASS_MULTICOLOR` is already activated for all targets except `stm32`, see target/linux/generic/config-6.6.  Just Activate it for `stm32` too.

hauke replied:
No need to select CONFIG_LEDS_CLASS_MULTICOLOR here any more.

namiltd commented:
No if only with led-group-multi-color. Yes, if it were to be used independently.

professor-jonny commented:
led-class-multicolor is enabled by default in kernel config as it is used by a few other drivers.
https://github.com/openwrt/openwrt/blob/main/target/linux/generic/config-6.6#L3046

namiltd commented:
In target stm32 # CONFIG_LEDS_CLASS_MULTICOLOR is not set

professor-jonny committed:
stm32: add default kernel config option CONFIG_LEDS_CLASS_MULTICOLOR

The generic target configuration activates the
CONFIG_LEDS_CLASS_MULTICOLOR option. Do not deactivate it for the stm32
target.

Signed-off-by: jonathan brophy <professor_jonny@hotmail.com>
Link: https://github.com/openwrt/openwrt/pull/17277
[Reword commit message]
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -274,7 +274,7 @@ CONFIG_KERNEL_GZIP=y
 CONFIG_KMAP_LOCAL_NON_LINEAR_PTE_ARRAY=y
 # CONFIG_LEDS_BRIGHTNESS_HW_CHANGED is not set
-# CONFIG_LEDS_CLASS_MULTICOLOR is not set
+# CONFIG_LEDS_CLASS_MULTICOLOR=Y
```
Just remove this line. If there is no special setting in this file it will take the value from the generic configuration.

hauke commented:
The CI complains about some formal problems, see https://github.com/openwrt/openwrt/actions/runs/12502659186/job/34881927584?pr=17277

hauke commented:
On x86_64 it fails with:
```
NOTICE: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/leds-gpio.ko' is built-in.
ERROR: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/rgb/leds-group-multicolor.ko' is missing.
make[2]: *** [modules/leds.mk:43: /__w/openwrt/openwrt/openwrt/bin/targets/x86/64/packages/kmod-led-group-multi-color-6.6.83-r1.apk] Error 1
time: package/kernel/linux/compile#19.02#16.00#30.76
Error: Process completed with exit code 2.
```

professor-jonny commented:
> On x86_64 it fails with:
> 
> ```
> NOTICE: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/leds-gpio.ko' is built-in.
> ERROR: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/rgb/leds-group-multicolor.ko' is missing.
> make[2]: *** [modules/leds.mk:43: /__w/openwrt/openwrt/openwrt/bin/targets/x86/64/packages/kmod-led-group-multi-color-6.6.83-r1.apk] Error 1
> time: package/kernel/linux/compile#19.02#16.00#30.76
> Error: Process completed with exit code 2.
> ```

I'm not really sure why but I suspect it does not compile in the x86 arch or is a path error not finding the leds/rgb/ directory with just that platform, ill try build with x86 and see what happens.

hauke commented on package/kernel/linux/modules/leds.mk:
```diff
@@ -24,6 +24,24 @@ $(eval $(call KernelPackage,leds-gpio))
+define KernelPackage/led-group-multi-color
+  SUBMENU:=$(LEDS_MENU)
+  TITLE:=LEDs group multi-color support
+  KCONFIG:=CONFIG_LEDS_CLASS_MULTICOLOR \
```
No need to select `CONFIG_LEDS_CLASS_MULTICOLOR` here it is already selected in `target/linux/generic/config-6.6`.

professor-jonny replied:
I removed that and it fixed building on the x86 platform I don't understand why.

hauke commented:
> > On x86_64 it fails with:
> > ```
> > NOTICE: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/leds-gpio.ko' is built-in.
> > ERROR: module '/__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/linux-x86_64/linux-6.6.83/drivers/leds/rgb/leds-group-multicolor.ko' is missing.
> > make[2]: *** [modules/leds.mk:43: /__w/openwrt/openwrt/openwrt/bin/targets/x86/64/packages/kmod-led-group-multi-color-6.6.83-r1.apk] Error 1
> > time: package/kernel/linux/compile#19.02#16.00#30.76
> > Error: Process completed with exit code 2.
> > ```
> 
> I'm not really sure why but I suspect it does not compile in the x86 arch or is a path error not finding the leds/rgb/ directory with just that platform, ill try build with x86 and see what happens.

`CONFIG_LEDS_GROUP_MULTICOLOR` depends on CONFIG_OF. x86 does not activate `CONFIG_OF`. When you remove `CONFIG_LEDS_CLASS_MULTICOLOR` from the Kconfig option it should only build the kmod when `CONFIG_LEDS_GROUP_MULTICOLOR` is set and the problem should be gone.

professor-jonny commented:
> `CONFIG_LEDS_GROUP_MULTICOLOR` depends on CONFIG_OF. x86 does not activate `CONFIG_OF`. When you remove `CONFIG_LEDS_CLASS_MULTICOLOR` from the Kconfig option it should only build the kmod when `CONFIG_LEDS_GROUP_MULTICOLOR` is set and the problem should be gone.

Do I need to enable `CONFIG_LEDS_CLASS_MULTICOLOR` in kconfig in the x86 platform or add it as a dependency or something? I don't really understand CONFIG_OF is this part of the menuconfig stuff so it is auto generated correct?
